### PR TITLE
[Attention] Add opt-in FlashInfer experimental FA2 paged-prefill integration

### PR DESCRIPTION
**Pull request by 0z5a**
**Incremental review:** [Files changed](https://github.com/vllm-project/vllm/pull/57709/files) against `main`.

## Scope

Related to https://github.com/flashinfer-ai/flashinfer/issues/5007 and the experimental implementation in https://github.com/flashinfer-ai/flashinfer/pull/4015. This PR is the **real vLLM opt-in integration** of the experimental unified paged-attention API — the gap the #4015 integration sketches left open (they were never run against a live engine).

In scope: FP16/BF16 Q/KV/output, SM89, native FA2, paged **prefill** only, ordinary causal attention, `lse_mode="none"`, behind an explicit opt-in. Default behavior is unchanged.
Out of scope: decode migration, cascade, DCP, attention sinks, logits soft cap, sliding window, spec-decode reorder, FP8/NVFP4 KV, default-backend changes. No part of #4015 is rewritten.

## Implementation

- `VLLM_FLASHINFER_PAGED_PREFILL` (bool, default `0`) and `VLLM_FLASHINFER_PAGED_PREFILL_WARMUP` (bool, default `1`) in `vllm/envs.py`.
- `PagedPrefill` metadata variant plus a `PagedPrefillAdapter` in `vllm/v1/attention/backends/flashinfer.py`; `build()` produces it for the prefill slice only, `forward()` runs it with `out=` pointing at the caller's pre-allocated prefill slice, and decode stays on the existing path untouched.
- The `Resolution` returned by `resolve_paged_attention()` is pinned once (at init when the KV layout is already resolved, otherwise at the first plan) and passed back to every `plan()`, so the opt-in cannot drift to another backend.
- `PagedAttentionMetadata.dense(...)` is built from the mirrors vLLM already owns (`qo_indptr_cpu`, `kv_seq_lens_cpu`) — no new hot-path D2H and no new per-forward synchronisation.
- With the opt-in off, the experimental module is not imported and no additional dependency is required. With it on but a prerequisite missing, the adapter **raises at start-up/build time** — it never silently falls back to the old path and then reports success. Rejections: non-SM89, non-fp16/bf16, quantized KV, DCP, attention sinks, logits soft cap, sliding window, spec-decode reorder, non-causal batches, batch-invariant mode, cascade, and a drifted/missing pinned API surface (the error names the required PR #4015 revision).
- Init-time warmup: because the first requests otherwise pay the one-time FlashInfer JIT/plan setup inside the measured window, the adapter issues one bounded synthetic plan+run (1 request, 1 query token, 1 page, throwaway zeros KV/query) right after the Resolution pin, and logs one INFO line with its duration. It loads the same FA2 module variant the hot path uses, runs before `/health` and before any capture, is skipped entirely when the opt-in is off, and adds nothing to `build()`/`forward()`.

## Validation

FlashInfer dependency: PR #4015 head `a05ffb0865a9a4b51609f09174cde4e423bfd290` on top of the pinned runtime `flashinfer-python==0.6.18`; vLLM base v0.29.0 (this branch is rebased onto current `main`).
Hardware and software: NVIDIA L20 (SM89), driver 595.91.07, CUDA 12.8, torch 2.13.0+cu130, Python 3.10.

Adapter suite (vLLM v0.29.0, where the engine measurements below were taken):

```
HF_HOME=<offline hf home> python -m pytest -q tests/v1/attention/test_flashinfer_paged_prefill_optin.py
# 14 passed, 0 failed at the pre-warmup revision; the warmup revision adds 5 cases (19 total)
```

On this branch (rebased onto current `main`) the same suite runs as **17 passed / 2 failed**. The two failures are the model-construction harness cases (`test_builder_routes_prefill_only_when_opted_in`, `test_warmup_runs_at_construction_and_nowhere_else`): both build a real `VllmConfig` from an HF model name, and `main`'s model inspection rejects the offline stand-in config this node has to use (`Model architectures ['LlamaForCausalLM'] failed to be inspected`) because huggingface.co is unreachable there. They pass on v0.29.0 with the same stand-in. This PR is therefore opened as a draft until those two cases have been run against a reachable model config.

Real-engine paired A/B on the L20 (`Qwen/Qwen3-0.6B`, `page_size=16`, `kv_layout=NHD`, one fresh server per arm, order A→P→P→A, 16 requests per workload at concurrency 4, greedy, prefix caching off):

| workload | metric | baseline (opt-in off) | opt-in on | delta |
|---|---|---|---|---|
| in=512/out=16, 4 warm client requests | TTFT mean | 125.4 ms | **61.9 ms** | −50.6% (2.03x) |
| in=512/out=16, 4 warm client requests | prefill throughput | 10 465 tok/s | **17 964 tok/s** | +71.6% |
| in=3072/out=16, 4 warm client requests | TTFT mean | 238.5 ms | 243.4 ms | −2.1% (within noise) |
| in=3072/out=16, 4 warm client requests | prefill throughput | 28 822 tok/s | **31 841 tok/s** | +10.5% |
| in=3072/out=16, 4 warm client requests | request wall mean | 423.4 ms | **371.5 ms** | +12.2% |
| **fresh server, no client warmup, warmup disabled** | prefill throughput | 12 876 / 36 136 tok/s | 9 582 / 20 892 tok/s | **−25.6% / −42.2%** |
| **fresh server, no client warmup, init warmup enabled** | prefill throughput | 11 779 / 43 325 tok/s | 12 539 / 41 386 tok/s | **+6.4% / −4.5%** |
| **fresh server, no client warmup, init warmup enabled** | TTFT mean | 114.2 / 163.5 ms | 104.4 / 169.0 ms | **+8.6% / −3.4%** |

Rows 6–8 are the point of the init warmup: measured from the very first request on a fresh server, the opt-in goes from a large regression to parity (a small gain on short prompts, a small loss on long ones, both within this shared node's noise). Row 6 is the same configuration with the warmup disabled, kept as the control. The warmup moves the one-time JIT/plan cost to start-up — `server.log` shows two warmup lines per start (the profiling and the real metadata builder), e.g. `FlashInfer paged prefill warmup (VLLM_FLASHINFER_PAGED_PREFILL=1): backend=fa2, kv_layout=NHD, page_size=16, batch=1x1, took 0.550 s`, before `/health` and before any capture.

It does **not** fully delete the cold path: the steady-state gains (rows 1–5) still need a couple of requests, so part of the first-request cost is not JIT and is not covered by this warmup. Per-plan/per-shape work stays per step by design.

Path evidence, not just the flag: with the opt-in on the server log contains `FlashInfer paged prefill uses the experimental unified API (VLLM_FLASHINFER_PAGED_PREFILL=1): backend=fa2, kv_layout=NHD, page_size=16` (and, with warmup, the warmup line before the first request); with it off neither line appears, and both arms log `Using AttentionBackendEnum.FLASHINFER backend.`

Independent API-level check (no vLLM): six paged-prefill configurations (single short, exact page, page−1, page+1, mixed lengths with a non-contiguous page table, multi-page) against an independent fp32 causal reference computed by gathering K/V from the page table — all six resolved `backend=fa2` with `max_abs_err` between 6.6e-3 and 8.4e-3.

NOT_RUN: TP2/TP4, CUDA graph/replan, over-capacity rejection, plan/metadata CPU time, peak-memory decomposition, and any model larger than the 0.6B used here.

## Results and limitations

Correctness: adapter suite green; the six engine-independent prefills agree with the independent reference; decode is untouched; the default path is unchanged with the opt-in off.
Performance: the table above; steady-state gains are concentrated in short-prompt prefill, and the long-prompt TTFT difference is inside the noise of this measurement.
Memory: not decomposed; the adapter adds ~144 MiB of FlashInfer scratch per metadata builder (128 MiB controller workspace + wrapper buffers), resident before the real KV tensors. The cudagraph memory-profiling window closes before builders are constructed, so the available-KV number should be unchanged — that claim still needs a live check.
Limitations: single GPU, eager, one small model, shared node (paired design mitigates but does not remove interference); no bootstrap interval on 16 requests per arm; sliding window is rejected even though FA2 could serve it (flagged rather than guessed).

Test and benchmark sources are retained locally; the validation results below refer to those local files.

### CODE REVIEW

**Comment by 0z5a** [comment]
`pre-commit` and `DCO` are green on this head (the missing sign-off is fixed). The remaining red check is `Check PR label and author merge count`, the first-time-contributor gate, which needs a maintainer label before the test lanes run. The PR stays a draft until the two model-construction cases under Validation can run against a reachable model config, as described there.
